### PR TITLE
Use ES6 modules by default

### LINKED ISSUE #159: Insert require/import statements at the top
It's vital for libraries being published also in es6 modules format that they do not have any `require` calls in their source code. In addition it should be slightly quicker in commonjs environment to have this cached once per file instead of loading for each prop type.

### PR DESCRIPTION
**Pull request by brigand**
WIP, resolves #159.

There are many edge cases where this might break, so I'd like someone to review my work before I continue. Any feedback is appreciated.

### CODE REVIEW

File: src/__tests__/__snapshots__/basic-test.js.snap [hunk 1]
```diff
@@ -75,10 +81,10 @@ Foo.propTypes = {
...
+  some_external_type: typeof (_types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func) === 'function' ? (_types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func).isRequired ? (_types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func).isRequired : _types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func : require('prop-types').shape(_types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func).isRequired,
   some_external_type_intersection: require('prop-types').shape(Object.assign({}, {
     foo: require('prop-types').string.isRequired
-  }, babelPluginFlowReactPropTypes_proptype_SomeExternalType === require('prop-types').any ? {} : babelPluginFlowReactPropTypes_proptype_SomeExternalType)).isRequired
+  }, (_types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func) === require('prop-types').any ? {} : _types.bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || _propTypes2.default.func)).isRequired
```
Comment (Andarist) [inline src/__tests__/__snapshots__/basic-test.js.snap]: some leftovers `require('prop-types')` are here, i guess its not intended

Comment (brigand) [reply src/__tests__/__snapshots__/basic-test.js.snap]: Good catch, thanks.

**Comment by brigand** [comment]
Had to use a hack [here] to avoid adding an argument to each function in `makePropTypesAst.js`.

[here]: https://github.com/brigand/babel-plugin-flow-react-proptypes/pull/162/commits/0a676d563c7f924b8892c79489b0dbaec141545a

Seems to be working in the general example.

```js
import { bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType } from './types';
import bpfrp_PropTypes from 'prop-types';
var babelPluginFlowReactPropTypes_proptype_Qux = {
  baz: bpfrp_PropTypes.oneOf(['literal']).isRequired
};
Foo.propTypes = {
  an_optional_string: bpfrp_PropTypes.string,
  a_number: bpfrp_PropTypes.number.isRequired,
  a_boolean: bpfrp_PropTypes.bool.isRequired,
  a_generic_object: bpfrp_PropTypes.object.isRequired,
  some_external_type: typeof (bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || bpfrp_PropTypes.func) === 'function' ? (bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || bpfrp_PropTypes.func).isRequired ? (bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || bpfrp_PropTypes.func).isRequired : bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || bpfrp_PropTypes.func : bpfrp_PropTypes.shape(bpfrp_babelPluginFlowReactPropTypes_proptype_SomeExternalType || bpfrp_PropTypes.func).isRequired,
```

**Comment by billyjanitsch** [comment]
This is awesome, @brigand. Thanks for working on this. ❤️

I'll need to spend some time digging into the source before I can usefully review your changes -- I'll try to find time this week.

**Comment by brigand** [comment]
It seems to be working on my main work project now. I also added some code to deal with circular dependencies. There should be more propTypes warnings due to that change (I had some).

I've made `useStatic` the default. It can be disabled with the `noStatic` option. 

Please try out the alpha.

```js
yarn add --dev babel-plugin-flow-react-proptypes@14.0.0-alpha.4
```

**Comment by brigand** [comment]
For anyone trying it, please try with `deadCode: true`  and `deadCode: false`.

**Comment by Andarist** [comment]
👋 any plans for releasing this? Is there anything left to be done? Do you need any help?

**Comment by brigand** [comment]
@Andarist have you tested the beta on a project? Ideally with ESM handled by webpack instead of babel.

**Comment by Andarist** [comment]
Not extensively, only on a rather small project (created especially for this endeavour). I plan to migrate to flow our ui kit at work but I'd need to have this new version of the plugin released first. Can start migrating soon-ish, but it will take some time before extensive tests could be made by me.

After a thought - I'll try to convert `react-virtualized` this week to use this alpha version.

**Comment by brigand** [comment]
@Andarist cool, I'll leave it up to you to give the go-ahead on merging/publishing this change.

**Comment by Andarist** [comment]
[This file](https://github.com/bvaughn/react-virtualized/blob/master/source/Grid/types.js) transpiled to wrapped cjs exports, something like:
```js
var babelPluginFlowReactPropTypes_proptype_CellPosition = {
  columnIndex: bpfrp_PropTypes.number.isRequired,
  rowIndex: bpfrp_PropTypes.number.isRequired
};
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_CellPosition', {
  value: babelPluginFlowReactPropTypes_proptype_CellPosition,
  configurable: true
});
```

The only thing that is in es format and that is probably a result of this plugin is:
```js
import bpfrp_PropTypes from 'prop-types';
```

**Comment by brigand** [comment]
@Andarist yeah, I think using an ES6 export there would be best.

New behavior, input:
```js
export type Foo = string | number;
```

Output (with babel modules transform disabled)

```js
var babelPluginFlowReactPropTypes_proptype_Foo = bpfrp_PropTypes.oneOfType([bpfrp_PropTypes.string, bpfrp_PropTypes.number]);
import bpfrp_PropTypes from "prop-types";
export { babelPluginFlowReactPropTypes_proptype_Foo };
```

**Comment by Andarist** [comment]
It's not yet published, right?

**Comment by brigand** [comment]
I think it's ready. Published as 14.0.0.
